app/helpers/pipeline_runs_helper.rb

```diff
@@ -91,6 +91,88 @@ def job_info(job_id, run_id)
+      sfn_execution_history_hash = JSON.parse(stdout)
+      sfn_hash = _parse_sfn_execution_history_hash(sfn_execution_history_hash)
+      job_status = sfn_hash[stage_number.to_s]["status"]
+    else
+      LogUtil.log_err_and_airbrake("Error for update sfn status for record #{run_id} - stage #{stage_number} with error #{stderr}")
+      job_status = PipelineRunStage::STATUS_ERROR # transient error, job is still "in progress"
+      job_status = PipelineRunStage::STATUS_FAILED if stderr =~ /ExecutionDoesNotExist/ # job no longer exists
+    end
+    [job_status, job_log_id, stdout]
+  end
+
+  def _parse_sfn_execution_history_hash(sfn_execution_history_hash)
+    failed_state =
+      sfn_execution_history_hash["events"]
+      .select { |evt| %w[ExecutionAborted ExecutionFailed ExecutionTimedOut].include?(evt["type"]) }
+      .first
+
+    name_regexp = /\A(.*)(SPOT|EC2|ReadOutput)\Z/
+    step_numbers = {
+      "HostFilter" => 1,
+      "NonHostAlignment" => 2,
+      "Postprocess" => 3,
+      "Experimental" => 4,
+    }
+    result =
+      sfn_execution_history_hash["events"]
+      .select { |evt| evt["type"] =~ /^TaskState/ }
+      .map do |evt|
+        details_key = evt.keys.find { |k| k =~ /EventDetails$/ }
+        {
+          "type" => evt["type"],
+          "name" => evt.dig(details_key, "name"),
+        }
+      end
+      .map do |evt|
+        stage, task = evt["name"].match(name_regexp)[1..2]
+        { "stage" => stage, "task" => task, "type" => evt["type"] }
+      end
+      .each_with_object({}) do |evt, acc|
+        stage, task, evt_type = evt.values_at("stage", "task", "type")
+        evt_obj = acc[stage] ||= evt.slice("stage")
+        task_obj = evt_obj[task] ||= {}
+        task_obj[evt_type] = true
+      end
+      .values
+      .map do |evt|
+        stage = evt["stage"]
+        started = evt.dig("SPOT", "TaskStateEntered") || evt.dig("EC2", "TaskStateEntered") || evt.dig("ReadOutput", "TaskStateEntered") || false
+        completed = evt.dig("ReadOutput", "TaskStateExited") || false
+        # I'm not using declared constants here because I'm emulating aws batch job `status` field
+        # trying to mimic output behavior of PipelineRunsHelper#job_info
+        # the domain for a job `status` field is SUBMITTED,PENDING,RUNNABLE,STARTING,RUNNING,SUCCEEDED,FAILED
+        status = if started && completed
+                   "SUCCEEDED"
+                 elsif started && !completed && !failed_state
+                   "RUNNING"
+                 elsif failed_state
+                   "FAILED"
+                 else
+                   "PENDING"
+                 end
+        [step_numbers[stage].to_s, { "stage" => stage, "status" => status }]
+      end
+      .to_h
+
+    if failed_state && !result["1"]
+      # special case when step function is aborted before stage 1
+      { "1" => { "stage" => "HostFilter", "status" => "FAILED" } }
+    else
+      result
+    end
+  end
```

Comment:
Prepending an identifier with an underscore with Ruby is most often used to indicate that something is unused, not private (like in Python). You can use the `private` keyword to make this method private.

In any case, I'm not sure why you're trying to parse the status of every stage when you only use the status for one of them. I wrote a function that just grabs the status of the `stage_number`, and it passes your tests:

```ruby
  def get_job_status(sfn_hash, stage)
    stage_numbers = {
      1 => "HostFilter",
      2 => "NonHostAlignment",
      3 => "Postprocess",
      4 => "Experimental",
    }

    task_status = {
      "TaskFailed" => "FAILED",
      "TaskSucceeded" => "SUCCEEDED"
    }

    status = "PENDING"
    last_status_id = 0

    # find exit task events
    exit_events = sfn_hash["events"].select { |event| event["type"] == "TaskStateExited" }

    # execution failed before any tasks were run
    if exit_events.length == 0
      if stage == 1
        status = "FAILED"
      end
    else
      # get the relevant exit events, and check the previous event, which will be success or failure
      stage_exit_events = exit_events.select { |event| event["stateExitedEventDetails"]["name"].start_with?(stage_numbers[stage]) }
      if stage_exit_events.length > 0
        last_stage_exit_event = stage_exit_events.last 
        last_status_event_index = last_stage_exit_event["previousEventId"] - 1
        last_status_event_type = sfn_hash["events"][last_status_event_index]["type"]
        status = task_status[last_status_event_type]
        last_status_id = last_stage_exit_event["id"]
      end
    end

    # check if there are events still running for that stage
    stage_entered_events = sfn_hash["events"].select { |event| event["type"] == "TaskStateEntered" && event["stateEnteredEventDetails"]["name"].start_with?(stage_numbers[stage])  }
    if stage_entered_events.length > 0 && stage_entered_events.last["id"] > last_status_id
      status = "RUNNING"
    end

    return status
  end
```

Reply:
About underscore: shouldn't be there. Removing it.
About not passing stage number: my reasoning for that was making it easier to test this method. Passing this stage and properly testing all conditions would add extra test conditions, one for each stage. It is worth pointing out this implementation will be thrown away soon.

Reply:
I agree it could be optimized, but I don't really think it worth investing time trying to do that, specially considering this is thrown away code. This code will be removed in future iterations, in favor of async results.

I appreciate your efforts to optimize that code, but I also believe may have overlooked some of the complexities of parsing those results. I tried to run your sample code against my tests (adapting my tests to send the stage id), and it looks like it is producing wrong results (you can see one example below).

Errors: `....FFFFFFFFFFFFFF.FF`
Example failures:
```
Failures:

  1) PipelineRunsHelper#get_succeeded_pipeline_run_ids_for_samples #_parse_sfn_execution_history_hash When processing sfn-execution-history_error-processing-output.json should eq {"1"=>{"stage"=>"HostFilter", "status"=>"FAILED"}}
     Failure/Error: it { is_expected.to eq(expected_result) }
     
       expected: {"1"=>{"stage"=>"HostFilter", "status"=>"FAILED"}}
            got: "RUNNING"
     
       (compared using ==)
     
       Diff:
       @@ -1,2 +1,2 @@
       -"1" => {"stage"=>"HostFilter", "status"=>"FAILED"},
       +"RUNNING"
       
     Shared Example Group: "check output" called from ./spec/helpers/pipeline_runs_helper_spec.rb:90
     # ./spec/helpers/pipeline_runs_helper_spec.rb:85:in `block (7 levels) in <top (required)>'
```

Reply:
I'm renaming the method with underscore to satisfy your review.

Just in the sake of curiosity, I scanned the rails library source code, and assuming rails is a good piece of code, it seems that Ruby community may not have such strong agreements in this regard. I could find tons of private methods in rails library that are prefixed with underscores, like this example:

https://github.com/rails/rails/blob/3269e4e7b04c06855464d48ed4f862a127abb5ba/actionview/lib/action_view/helpers/url_helper.rb#L50